### PR TITLE
Add Protobuf protoc 3.9.1

### PR DESCRIPTION
**Pull request by uilianries**
Specify library name and version:  **protoc/3.9.1**

This is the second part of Google Protobuf. It requires #64 first!

Also, Protoc provides CMake with specific macros, which can't be exported by Conan, so this package requires an exception in the whitelist.

I've added the path to builddirs as required by KB-H019.

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by uilianries** [comment]
We need to put protoc in the whitelist, since the macros are declared in protoc-config.cmake

**Comment by lasote** [comment]
> We need to put protoc in the whitelist, since the macros are declared in protoc-config.cmake

No, that cannot be whitelisted. We need a mechanism to declare macros to be inherited: https://github.com/conan-io/conan/issues/5798

**Comment by uilianries** [comment]
Could we add a new tag for this PR, like `blocked` ?

**Comment by danimtb** [comment]
Yeah, it is reasonable

File: recipes/protoc/3.9.x/metadata.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"repo": {"type": "git", "url": "git@github.com:uilianries/conan-center-index.git", "revision": "375cc841b47ed6da559ec7330fb5a90d7bbdd5e9", "dirty": false}}
```
[the remark this answers is not in the document]
Comment (uilianries) [reply recipes/protoc/3.9.x/metadata.json]: I really need to add .gitignore for it! Thanks!

**Comment by uilianries** [comment]
@danimtb check my last [commit](https://github.com/conan-io/conan-center-index/pull/65/commits/d9dc4373825610f2c3b3592f66eabac6d1cc7259).

The Conan generator `cmake_find_package` does not append env vars (e.g. PATH), which were exported by `self.env_info`, this behavior is a problem for those macros exported by Protobuf files. As the macro don't know where protoc is installed, it will fail, so we need to use the regular `cmake` generator and patch the protobuf cmake file. 

As you can see, it's fragile, and don't like to force anyone using the `cmake` generator, when `cmake_find_package` should enough. I can't find a smart solution for this situation, I only can think about changing the `cmake_find_package` when building an installer package.

**Comment by Croydon** [comment]
Do we have any kind of plan how to progress with Protobuf/Protoc? 🤔

**Comment by madebr** [comment]
@uilianries Check https://github.com/uilianries/conan-center-index/pull/1

File: recipes/protoc/3.9.x/test_package/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+if (NOT protobuf_MODULE_COMPATIBLE)
+    message(FATAL_ERROR "protobuf_MODULE_COMPATIBLE is mandatory")
+endif()
+
+add_executable(${PROJECT_NAME} test_package.cpp addressbook.proto)
+protobuf_generate_cpp(PROTO_SRCS PROTO_HDRS TARGET ${PROJECT_NAME})
+message(STATUS "PROTO_SRCS: ${PROTO_SRCS}")
+message(STATUS "PROTO_HDRS: ${PROTO_HDRS}")
+target_include_directories(${PROJECT_NAME} PRIVATE ${PROJECT_BINARY_DIR})
+target_link_libraries(${PROJECT_NAME} PRIVATE Protobuf::Protobuf)
```
Comment (madebr) [inline recipes/protoc/3.9.x/test_package/CMakeLists.txt]:
```suggestion
target_link_libraries(${PROJECT_NAME} PRIVATE Protobuf::Protobuf)
set_property(TARGET ${PROJECT_NAME} PROPERTY CXX_STANDARD 11)
```

**Comment by madebr** [comment]
@uilianries 
Do you have an idea?
```
CMake Error at conanbuildinfo.cmake:461 (message):
  Detected a mismatch for the compiler version between your conan profile
  settings and CMake:

  Compiler version specified in your conan profile: 14

  Compiler version detected in CMake: 19.24

  Please check your conan profile settings (conan profile show
```

**Comment by Croydon** [comment]
Coming from https://github.com/conan-io/conan-center-index/issues/1136

we might want to add the protoc binaries directly to the protobuf package now

**Comment by uilianries** [comment]
@Croydon thanks for pinging!

**Comment by madebr** [comment]
FYI.
I'm working on it atm.
I need it for a recipe.

**Comment by madebr** [comment]
Looking at the release page of protobuf, they release protobuf and protoc binaries seperately.
See https://github.com/protocolbuffers/protobuf/releases

**Comment by uilianries** [comment]
I gonna clone this PR since we are moving to a new approach

**Comment by 0x8000-0000** [comment]
@uilianries - can you put here a link to the new approach so we can follow at home :)?

Thanks!

**Comment by uilianries** [comment]
@0x8000-0000 https://github.com/conan-io/conan-center-index/pull/1179

**Pull request closed without merging**